lib/instrumentation/modules/kafkajs.js

```diff
@@ -0,0 +1,451 @@
+      }
+
+      // We do not add headers or body because:
+      // - `record.messages` is a list
+      // - spec says is for transactions (https://github.com/elastic/apm/blob/main/specs/agents/tracing-instrumentation-messaging.md#transaction-context-fields)
+      span.setMessageContext({ queue: { name: topic } });
+
+      const service = {
+        resource: `${SUBTYPE}/${topic}`,
+        type: SUBTYPE,
+        name: topic,
+      };
+
+      span._setDestinationContext({ service });
+
+      let result, err;
+      try {
+        result = await origSend.apply(this, arguments);
+      } catch (ex) {
+        // Save the error for use in `finally` below, but re-throw it to
+        // not impact code flow.
+        err = ex;
+        throw ex;
+      } finally {
+        span.setOutcome(
+          err ? constants.OUTCOME_FAILURE : constants.OUTCOME_SUCCESS,
+        );
+        span.end();
+      }
+
+      return result;
+    };
+  }
+
+  /**
+   * Returns the patched version of `Producer.sendBatch` which
+   * - creates an exit span for the operation
+   * - propagates trace context via message headers
+   *
+   * @param {Producer['sendBatch']} origSendBatch
+   * @returns {Producer['sendBatch']}
+   */
+  function wrapProducerSendBatch(origSendBatch) {
+    return async function (batch) {
+      // TODO: discuss this. if all topics are ignored should the agent avoid
+      // creating a span?
+      let span;
+      const topics =
+        batch.topicMessages && batch.topicMessages.map((tm) => tm.topic);
+      const shouldIgnoreBatch =
+        topics && topics.every((t) => shouldIgnoreTopic(t, config));
+
+      if (!shouldIgnoreBatch) {
+        span = ins.createSpan(
+          `${NAME} send messages batch`,
+          TYPE,
+          SUBTYPE,
+          'send',
+          { exitSpan: true },
+        );
+      }
+
+      // W3C trace-context propagation.
+      const runContext = ins.currRunContext();
+      const parentSpan =
+        span || runContext.currSpan() || runContext.currTransaction();
+
+      if (parentSpan && batch.topicMessages) {
+        batch.topicMessages.forEach((topicMessage) => {
+          topicMessage.messages.forEach((msg) => {
+            const newHeaders = Object.assign({}, msg.headers);
+            parentSpan.propagateTraceContextHeaders(
+              newHeaders,
+              function (carrier, name, value) {
+                // TODO: why doing it everywhere???
+                if (name.startsWith('elastic-')) {
+                  return;
+                }
+                carrier[name] = value;
+              },
+            );
+            msg.headers = newHeaders;
+          });
+        });
+      }
+
+      if (!span) {
+        return origSendBatch.apply(this, arguments);
+      }
+
+      const destCtx = {
+        // `service.name` and `service.resource` cannot be set since `producer.sendBatch`
+        // is meant to send to several topics at a time
```

Comment:
This might be worth discussing with the team. What does the Java Kafka instr do in this case? If there is a single topic in the outgoing batch, then should this use that value?

Reply:
I can check the Java agent Implementation 1st before raising this topic in the next meeting.